### ISSUE TITLE
[HOLD for VSB] [$250] Public room - Fallback avatars displayed/avatars don't load when scrolling up as anon user

### ISSUE THREAD

**Issue by lanitochka17** [Expensify/App#43423]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

**Version Number:** 1.4.81-4
**Reproducible in staging?:** Y
**Reproducible in production?:** Y
**If this was caught during regression testing, add the test name, ID and link from TestRail:** https://expensify.testrail.io/index.php?/tests/view/4610474
**Email or phone of affected tester (no customers):** 
**Issue reported by:** Applause - Internal Team

## Action Performed:
1. Navigate to a public room with long existing chat history, e. g. https://staging.new.expensify.com/r/2652281908731856
as anon user
2. Scroll to the top of the conversation

## Expected Result:
Avatars of users should load along with the chat history

## Actual Result:
Avatars of users do not load, fallback avatars are displayed instead
Avatars do not load even after a few minutes waiting time

## Workaround:
Unknown

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [ ] Android: Native
- [ ] Android: mWeb Chrome
- [ ] iOS: Native
- [ ] iOS: mWeb Safari
- [x] MacOS: Chrome / Safari
- [ ] MacOS: Desktop

## Screenshots/Videos

Add any screenshot/video evidence

</details> 

https://github.com/Expensify/App/assets/78819774/5d548520-ec77-4650-9806-d870a6778ccc

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~01897ff6965422a788</li>
        <li>Upwork Job ID: 1800422322383265958</li>
        <li>Last Price Increase: 2024-06-11</li>
    </ul>
</details>

<details><summary>Issue Owner</summary>Current Issue Owner: @Christinadobrzyn </details>

**Comment by lanitochka17** [comment]
@Christinadobrzyn  FYI I haven't added the External label as I wasn't 100% sure about this issue. Please take a look and add the label if you agree it's a bug and can be handled by external contributors

**Comment by lanitochka17** [comment]
We think that this bug might be related to #vip-vsp

**Comment by Christinadobrzyn** [comment]
I can reproduce based on the steps in the OP - I think this can be external

**Comment by truph01** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
Avatars of users do not load, fallback avatars are displayed instead
Avatars do not load even after a few minutes waiting time

### What is the root cause of that problem?
In a public room there're a lot of actors because a lot of people send messages there.

When the report is initially loaded, it will only load a subset of the latest report actions, along with the personal details for the actors of those report actions.

When the user scrolls up, older report actions are loaded by `GetOlderActions`, but the personalDetails of the actors of those older actions are still lacking.

So there's no user avatar to show and there'll be the disabled cursor from this line https://github.com/Expensify/App/blob/65db65069390cdd9fd386e087b6d8eed94ee22b0/src/pages/home/report/ReportActionItemSingle.tsx#L172

### What changes do you think we should make in order to solve the problem?

When we try to open the profile of a user from deep link, and that profile doesn't exist in personal details, we'll [call openPublicProfilePage](https://github.com/Expensify/App/blob/65db65069390cdd9fd386e087b6d8eed94ee22b0/src/pages/ProfilePage.tsx#L162) to fetch more info on those users. Likewise for [calling `openReport`](https://github.com/Expensify/App/blob/65db65069390cdd9fd386e087b6d8eed94ee22b0/src/pages/home/report/withReportOrNotFound.tsx#L68) when navigating to a report route that doesn't exist in Onyx data.

So for this case of personal details missing from `Onyx`, if the case is detected, we should also call an endpoint to fetch data on the users. We can use `openPublicProfilePage`.

More ideally, if there's a back-end endpoints that allow fetching multiple public profile at once, something like `openPublicProfiles`, we can batch the `accountIds` that are missing personal details (for instance in a 3 seconds window) and fetch them in 1 API call.

### What alternative solutions did you explore? (Optional)

**Comment by TheGithubDev** [comment]
### **Proposal**

### **Please re-state the problem that we are trying to solve in this issue:**
The problem is that avatars of users do not load when an anonymous (anon) user scrolls up in a public room with a long chat history. Instead, fallback avatars are displayed and remain even after waiting for a few minutes.

### **What is the root cause of that problem?**
The root cause of this problem is likely due to issues with loading user avatars for anonymous users when fetching older messages in a chat room. This could be caused by:
1. **Network Requests**: The network requests for fetching user avatars might not be triggered correctly or might fail when the user is anonymous.
2. **Caching:** There might be issues with caching that prevent avatars from being loaded correctly for anonymous users.
3. **Permission:** Anonymous users might not have the proper permissions to load user avatars.

### **What changes do you think we should make in order to solve the problem?**
To address this issue, we should:
1. **Investigate Network Requests:**
- Ensure that the network requests to load user avatars are being made correctly and handle failures gracefully.
2. **Adjust Caching Mechanisms:**
- Review and adjust the caching mechanisms to ensure avatars can be fetched correctly even for anonymous users.
3. **Ensure Proper Permissions:**
- Check and update permissions to make sure anonymous users can access user avatars.
4. **Update Fallback Logic:**
- Improve the fallback logic to retry fetching avatars or show a loading indicator instead of static fallback avatars.

**Example of changes:**

1. **Network Requests:**
- Ensure avatar loading requests are being sent when scrolling up as an anonymous user:
   ```javascript
   useEffect(() => {
       if (isAnonUser && isFetchingOlderMessages) {
           fetchUserAvatars();
       }
   }, [isAnonUser, isFetchingOlderMessages]);
   ```

2. **Caching Mechanisms:**
- Adjust caching mechanisms to ensure avatars are fetched correctly:
   ```javascript
   if (cache.hasAvatar(userId)) {
       displayAvatar(cache.getAvatar(userId));
   } else {
       fetchAvatar(userId)
           .then(avatar => {
               cache.setAvatar(userId, avatar);
               displayAvatar(avatar);
           })
           .catch(() => {
               displayFallbackAvatar();
           });
   }
   ```

3. **Permissions:**
- Ensure anonymous users have the necessary permissions to load avatars:
   ```javascript
   if (isAnonUser) {
       ensureAvatarPermissions(userId);
   }
   ```

4. **Fallback Logic:**
- Improve fallback logic to retry fetching avatars or show a loading indicator:
   ```javascript
   if (!avatarLoaded) {
       displayLoadingIndicator();
       fetchAvatar(userId)
           .then(avatar => {
               displayAvatar(avatar);
           })
           .catch(() => {
               displayFallbackAvatar();
           });
   }
   ```

### **What alternative solutions did you explore?**

**1. Use Default Avatars**: 
- Instead of attempting to load the actual avatars for anonymous users, use default avatars. This approach would ensure a consistent user experience but might lack the personalization of user avatars.

**2. Progressive Loading**: 
- Implement a progressive loading strategy where low-resolution avatars load first and are replaced by high-resolution versions as they become available. This could improve the perceived 
performance and user experience.

**3. Batch Loading:**
- Load avatars in batches to reduce the number of network requests and improve performance. This approach would require managing the state of which avatars have been requested and loaded.

_Would be glad to get this fixed asap_
_**#the GitHub Dev**_

**Comment by Christinadobrzyn** [comment]
Just a heads up - I'm going to be ooo until June 24th so going to assign a teammate to watch this while I'm away

@kadiealexander we are collecting proposals 

@sobitneupane would you be able to review the proposals we have?

**Comment by sobitneupane** [comment]
Thanks for your [proposal](https://github.com/Expensify/App/issues/43423#issuecomment-2160015782) @truph01 

Overall your proposal looks good to me. But I am concerned about the performance as we will be making api request for each `ReportActionItem` without personal details. Will it cause any impact on performance?

**Comment by truph01** [comment]
> Will it cause any impact on performance?

@sobitneupane From my testing, I don't see any performance impact.

There likely won't be performance issues because:
- For each personal detail that's missing, we'll call the API only once (if the personal details is already `isLoading`, we won't call)
- Each time the user scrolls up and fetch older actions, there'll like 6-7 new distinct actorAccountIds, this is low and shouldn't cause issues because the `GET` API is fast and the API calls will be sequential

**Comment by sobitneupane** [comment]
Thanks for the update @truph01 

[The proposal](https://github.com/Expensify/App/issues/43423#issuecomment-2160015782) from @truph01 looks good to me.

🎀 👀 🎀 C+ reviewed

**Comment by cristipaval** [comment]
I'm not happy about calling an API for each missing personal detail. We should fix the backend and make it return the personal details properly when loading a new batch of report actions with the `GetOlderActions` command. A new API command just for fetching a batch of missing personal details might be against our 1:1:1 pattern, which requires the App to make one single API call for a user action. Scroll Up user action should trigger the `GetOlderActions` command only.

**Comment by cristipaval** [comment]
I made this one `Internal` for now. Next week, I'll investigate the implementation of the `GetOlderActions` command and come back to this issue.

**Comment by cristipaval** [comment]
I'm afraid I won't be able to take care of this one this week. I have my plate full and am also assigned to the API monitoring this week.

**Comment by Christinadobrzyn** [comment]
Sounds good @cristipaval! I will add a hot pick label to this to see if we can get a volunteer

**Comment by Christinadobrzyn** [comment]
still looking for a volunteer

**Comment by Christinadobrzyn** [comment]
still looking for a volutneer!

**Comment by Christinadobrzyn** [comment]
Nothing yet, looking for a volunteer. I'll retest this tomorrow to see if anything has changed.

**Comment by Christinadobrzyn** [comment]
@cristipaval do you think you'd be able to work on this or should I reach out in the VSB channel for a volunteer?

**Comment by cristipaval** [comment]
unfortunately I'm not able, I need to also hand over some work from my plate given that I am preparing for a long leave

**Comment by Christinadobrzyn** [comment]
No worries! I asked in the VSB room if someone can grab this - https://expensify.slack.com/archives/C066HJM2CAZ/p1721930317057559

**Comment by Christinadobrzyn** [comment]
no takers yet, we'll waiting for a volunteer

**Comment by Christinadobrzyn** [comment]
nothing - I think this VSB is on hold, I'm going to move this out to weekly while we wait for a volunteer.

**Comment by Christinadobrzyn** [comment]
no volunteers but VSB is still on hold. Added a HOLD to the title and moving this to monthly.

**Comment by Christinadobrzyn** [comment]
no takers yet, I'll retest when I'm back from ooo on 9/10

**Comment by Christinadobrzyn** [comment]
testing this again, I think this is fixed. Asking QA to test again. https://expensify.slack.com/archives/C9YU7BX5M/p1726072341148429

**Comment by kavimuru** [comment]
Issue is reproducible.

https://github.com/user-attachments/assets/cdd42d97-4bbc-4b64-a203-bb66040cbfdc

**Comment by Christinadobrzyn** [comment]
Ah okay, so it looks like this is still happening with very long threads. We're still waiting for this to get picked up

**Comment by Christinadobrzyn** [comment]
Asking the team what we should be doing with this GH since vsb is on hold - https://expensify.slack.com/archives/C066HJM2CAZ/p1726152950736669

**Comment by Christinadobrzyn** [comment]
I'm going to close this given that vip-vsb is cancelled and this relates solely to that project.
